### ISSUE TITLE
Need to relax upper bound on base

### ISSUE THREAD

**Issue by svenpanne** [tibbe/hashable#107]
GHC's head is now at base-4.9.0.0, so the upper bound has to be relaxed, and a new version is needed on Hackage.

**Comment by hvr** [comment]
@svenpanne Generally I'm not sure if that's advisable at this point to start pushing `base`-relaxed packages to Hackage already, as we're not even in RC-phase for GHC 8

Don't forget you can easily add `--allow-newer=base` to your travis jobs for testing GHC HEAD, if it's just for the `base` version number

/cc @ekmett

**Comment by svenpanne** [comment]
As it is today, I basically can't test anything at all with HEAD without heavy hand-tweaking because of broken dependencies, so I actually think it's a good idea to let people know as early as possible that their packages need an update. If we would wait until GHC 8.0 is final, we would probably have tons of broken packages on Hackage for months.

Actually this kind of breakage is the reason why I'm personally opposed to upper bounds. I'm well aware of the trouble when they're not there, but I'm not fully convinced that the current trouble is less important, but that's another story...

**Comment by hvr** [comment]
The risk of uploading packages with a relaxed bound accepting a non-released `base` is that we may need to fix it up lateron, if it turns out that `base` changed more than expected. At least now we're in a position to be able to fixup broken bounds after they got uploaded to Hackage already via `.cabal` editing. Actually, we can just as well just relax the bounds for the existing hashable release, if it happens to still build w/ GHC HEAD.

**Comment by svenpanne** [comment]
What's the current state of this? Things are still broken after month, can we at least relax the version bound on Hackage?

**Comment by hvr** [comment]
Turns out, this is actually a good thing that all `hashable` releases reject base-4.9, as we'd need this anyway due to https://github.com/ekmett/semigroups/issues/56

I.e. we need to a new `hashable` release which takes into account NonEmpty before we can support `base-4.9`.

**Comment by svenpanne** [comment]
OK, I wasn't aware of that issue, I just wanted to make sure that the problem with `hashable`  has not been forgotten. Any ETA?

**Comment by tibbe** [comment]
As soon as base-4.9 is released, so I know what's in there, I'll make a release.

**Comment by hvr** [comment]
@tibbe btw, I just sent you an email regarding two possible ways forward to this

**Comment by svenpanne** [comment]
I'm a bit unhappy with the whole process: Releasing only when base-4.9 is done is too late IMHO. Giving library maintainers a way to test their stuff before GHC 8.0RC1 is release would be much better, otherwise tons of stuff is broken on that release day (and will probably stay like for some time this due to christmas). I understand that it would be a bit annoying if a something has to be re-released, but I don't see a way around that.

**Comment by tibbe** [comment]
I understand the frustration @svenpanne, but early releases both create extra work (i.e. in this case I would have had to make another release, try to blacklist the prior releases, and so forth). I think it will have to be expected that it takes a few weeks before a new GHC release is usable.

You're still able to test against a new hashable. Just use something like stack to build it from source. You can then use the HEAD version, where we can make the needed changes without releasing them.

**Comment by svenpanne** [comment]
Manual testing with e.g. stack is definitely not an option IMHO, I rely heavily on Travis CI's build matrices to support a wide range of GHC releases. I definitely won't put any more effort into this, it is already enough frustrating work. If we want to move fast and break things repeatedly (that seems to be the consensus on the mailing lists, and that might even be OK), we should at least make it easy for maintainers. And currently that's not the case at all...

**Comment by svenpanne** [comment]
Any progress here? I think most of the changes for base-4.9 are already committed and it would be nice to continue testing via Travis CI.

**Comment by hvr** [comment]
(Somewhat off-topic, but I wonder if Hackage (and cabal) should extend its candidate feature to support unofficial package releases as a volatile opt-in software-channel for this purpose. This is something that may become easier with the "local package" support that's being added to Cabal)

**Comment by ezyang** [comment]
I have to agree with @svenpanne, at the *very least*, hashable HEAD should be tracking base HEAD, so that if you are going to manually build, you can clone, make it visible and it will work. (Right now that's not the case; the base bound is still not updated.)

@svenpanne, the Cabal nix-local-build has made it even easier to add a vendored version of a library like hashable in case Hackage's version is not updated enough. When it gets into Cabal proper (which we're hoping to do as a tech preview), hopefully this will make the Travis situation easier. (Unfortunately, I've never worked with a Travis matrix so I don't entirely know what the moving parts here are.)

**Comment by svenpanne** [comment]
@ezyang: Setting up a build matrix on Travis CI is a bit verbose due to some restrictions in the YAML format, but essentially very easy, see e.g. [GLUT's Travis CI config](https://github.com/haskell-opengl/GLUT/blob/master/.travis.yml), which results in the following [matrix](https://travis-ci.org/haskell-opengl/GLUT). I'm relying on Herbert's PPA for GHC/cabal packages, which is very convenient.

As you can see from the config, I explicitly checkout/compile/install my own packages which I depend on, so I can see if the stuff in my own repos is consistent. This is just one side of the medal, the other one is: Is the stuff released on Hackage consistent, but that is already automatically checked by the [stackage bots](https://www.stackage.org/package/GLUT).

It would be great if upcoming cabal releases supported both use cases in a convenient way, doing all this cloning/building is a bit verbose and not really elegant. Perhaps there is already an easier way, but I don't know one.
